source/slang/slang-ir-diff-jvp.cpp

```diff
@@ -3,10 +3,303 @@
 
 #include "slang-ir.h"
 #include "slang-ir-insts.h"
+#include "slang-ir-clone.h"
 
 namespace Slang
 {
 
+struct JVPTranscriber
+{
+
+    // Stores the mapping of arbitrary 'R-value' instructions to instructions that represent
+    // their differential values.
+    Dictionary<IRInst*, IRInst*> instMapD;
+
+    // Cloning environment to hold mapping from old to new copies for the primal
+    // instructions.
+    IRCloneEnv cloneEnv;
+
+    void mapDifferentialInst(IRInst* instP, IRInst* instD)
+    {
+        instMapD.Add(instP, instD);
+    }
+
+    IRInst* getDifferentialInst(IRInst* instP)
+    {
+        return instMapD[instP];
+    }
+
+    IRFuncType* differentiateFunctionType(IRBuilder* builder, IRFuncType* funcType)
+    {
+        List<IRType*> parameterTypesD;
+        IRType* returnTypeD;
+
+        // Add all primal parameters to the list.
+        for (UIndex i = 0; i < funcType->getParamCount(); i++)
+        {   
+            parameterTypesD.add(funcType->getParamType(i));
+        }
+
+        // Add differential versions for the types we support.
+        for (UIndex i = 0; i < funcType->getParamCount(); i++)
+        {   
+            if (auto typeD = differentiateType(builder, funcType->getParamType(i)))
+                parameterTypesD.add(typeD);
+        }
+
+        // Transcribe return type. 
+        // This will be void if the primal return type is non-differentiable.
+        //
+        returnTypeD = differentiateType(builder, funcType->getResultType());
+        if (!returnTypeD)
+            returnTypeD = builder->getVoidType();
+
+        return builder->getFuncType(parameterTypesD, returnTypeD);
+    }
+
+    IRType* differentiateType(IRBuilder* builder, IRType* typeP)
+    {
+        switch (typeP->getOp())
+        {
+            case kIROp_HalfType:
+            case kIROp_FloatType:
+            case kIROp_DoubleType:
+                return builder->getType(typeP->getOp());
+            
+            default:
+                return nullptr;
+        }
+    }
+    
+    IRInst* differentiateParam(IRBuilder* builder, IRParam* paramP)
+    {
+        if (IRType* typeD = differentiateType(builder, paramP->getFullType()))
+        {
+            IRParam* paramD = builder->emitParam(typeD);
+            SLANG_ASSERT(paramD);
+            return paramD;
+        }
+        return nullptr;
+    }
+
+    List<IRParam*> transcribeParams(IRBuilder* builder, IRInstList<IRParam> paramListP)
+    {
+        // Clone (and emit) all the primal parameters.
+        List<IRParam*> newParamListP;
+        for (auto paramP : paramListP)
+        {
+            newParamListP.add(as<IRParam>(cloneInst(&cloneEnv, builder, paramP)));
+        }
+
+        // Now emit differentials.
+        List<IRParam*> newParamListD;
+        for (auto paramP : newParamListP)
+        {
+            IRParam* paramD = as<IRParam>(differentiateParam(builder, paramP));
+            mapDifferentialInst(paramP, paramD);
```

Comment:
We can add a IRNameHintDecoration here with `IRBuilder::addNameHintDecoration` on the param if we want to see a more explanatory parameter name in the generated code.